File: app/MyVentoryApp/lib/Model/lending.dart
import '../Model/user.dart';
import '../Model/item.dart';

class Lending {
  final int transactionId;
  final int? borrowerId;
  final String borrowerName;
  final String? borrowerEmail;
  final int lenderId;
  final String lenderName;
  final String? lenderEmail;
  final DateTime dueDate;
  final DateTime lendingDate;
  final List<ItemLending>? lendItems;
  final UserAccount? borrower;
  final UserAccount lender;
  final bool? isFinished;

  Lending({
    required this.transactionId,
    this.borrowerId,
    required this.borrowerName,
    this.borrowerEmail,
    required this.lenderId,
    required this.lenderName,
    this.lenderEmail,
    required this.dueDate,
    required this.lendingDate,
    this.lendItems,
    this.borrower,
    required this.lender,
    this.isFinished,
  });

  factory Lending.fromJson(Map<String, dynamic> json) {
    return Lending(
      transactionId: json["transactionId"],
      borrowerId: json["borrowerId"],
      borrowerName: json["borrowerName"],
      borrowerEmail: json["borrowerEmail"],
      lenderId: json["lenderId"],
      lenderName: json["lenderName"],
      lenderEmail: json["lenderEmail"],
      dueDate: DateTime.parse(json["dueDate"]),
      lendingDate: DateTime.parse(json["lendingDate"]),
      lendItems: json["lendItems"] != null
          ? (json["lendItems"] as List<dynamic>)
              .map((item) => ItemLending.fromJson(item))
              .toList()
          : null,
      borrower: json["borrower"] != null
          ? UserAccount.fromJson(json["borrower"])
          : null,
      lender: UserAccount.fromJson(json["lender"]),
      isFinished: json["isFinished"] ?? false,
    );
  }

  Lending copyWith({
    int? transactionId,
    int? borrowerId,
    String? borrowerName,
    String? borrowerEmail,
    int? lenderId,
    String? lenderName,
    String? lenderEmail,
    DateTime? dueDate,
    DateTime? lendingDate,
    List<ItemLending>? lendItems,
    UserAccount? borrower,
    UserAccount? lender,
    bool? isFinished,
  }) {
    return Lending(
      transactionId: transactionId ?? this.transactionId,
      borrowerId: borrowerId ?? this.borrowerId,
      borrowerName: borrowerName ?? this.borrowerName,
      borrowerEmail: borrowerEmail ?? this.borrowerEmail,
      lenderId: lenderId ?? this.lenderId,
      lenderName: lenderName ?? this.lenderName,
      lenderEmail: lenderEmail ?? this.lenderEmail,
      dueDate: dueDate ?? this.dueDate,
      lendingDate: lendingDate ?? this.lendingDate,
      lendItems: lendItems ?? this.lendItems,
      borrower: borrower ?? this.borrower,
      lender: lender ?? this.lender,
      isFinished: isFinished ?? this.isFinished,
    );
  }
}

class ItemLending {
  final int transactionId;
  final int itemId;
  final int quantity;
  final Lending lending;
  final InventoryItem item;

  ItemLending({
    required this.transactionId,
    required this.itemId,
    required this.quantity,
    required this.lending,
    required this.item,
  });

  // Factory constructor to create an ItemLending object from JSON
  factory ItemLending.fromJson(Map<String, dynamic> json) {
    return ItemLending(
      transactionId: json["transactionId"],
      itemId: json["itemId"],
      quantity: json["quantity"],
      lending: Lending.fromJson(json["lending"]),
      item: InventoryItem.fromJson(json["item"]),
    );
  }
}

File: backend/MyVentoryApi/Models/Lending.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Builders;

namespace MyVentoryApi.Models;

[Table("Lendings")]
public class Lending
{
    /* Database Table entries */
    [Key]
    [Required]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int TransactionId { get; set; }
    [ForeignKey("User")]
    public int? BorrowerId { get; set; } // Not required because the borrower can be a guest not registered in the system
    [MaxLength(50)]
    public string? BorrowerName { get; set; } // Not required because the borrower can be a user, identified by its UserID
    [Required]
    [ForeignKey("User")]
    public int LenderId { get; set; }
    [Required]
    public DateTime DueDate { get; set; }
    [Required]
    public DateTime LendingDate { get; set; }
    public DateTime? ReturnDate { get; set; }

    /* Navigation Properties */
    public virtual ICollection<ItemLending> LendItems { get; set; }
    public virtual User? Borrower { get; set; }
    public virtual User Lender { get; set; }

    public Lending(User borrower, User lender, DateTime dueDate)
    {
        BorrowerId = borrower.Id;
        BorrowerName = null;
        LenderId = lender.Id;
        DueDate = dueDate;
        LendingDate = DateTime.UtcNow;

        LendItems = new HashSet<ItemLending>();
        Borrower = borrower;
        Lender = lender;
        ReturnDate = null;
    }
    // Overloaded constructor for when the borrower is not a registered user
    public Lending(string borrowerName, User lender, DateTime dueDate)
    {
        BorrowerId = null;
        BorrowerName = borrowerName;
        LenderId = lender.Id;
        DueDate = dueDate;
        LendingDate = DateTime.UtcNow;

        LendItems = new HashSet<ItemLending>();
        Borrower = null;
        Lender = lender;

    }
    // Parameterless constructor for EF
    public Lending() : this(
        new User(),
        new User(),
        DateTime.UtcNow)
    { }
}

public class LendingConfiguration : IEntityTypeConfiguration<Lending>
{
    public void Configure(EntityTypeBuilder<Lending> builder)
    {
        /* Configure keys*/

        /* Configure other properties */
        builder.ToTable("Lendings");

        /* Configure navigation properties */
        builder.HasOne(l => l.Borrower)
               .WithMany(u => u.BorrowedItems)
               .HasForeignKey(l => l.BorrowerId)
               .IsRequired(false)
               .OnDelete(DeleteBehavior.NoAction);

        builder.HasOne(l => l.Lender)
               .WithMany(u => u.LentItems)
               .HasForeignKey(l => l.LenderId)
               .OnDelete(DeleteBehavior.NoAction);

        builder.HasMany(l => l.LendItems)
               .WithOne(il => il.Lending)
               .HasForeignKey(il => il.TransactionId)
               .OnDelete(DeleteBehavior.NoAction);
    }
}
